### PR TITLE
fix(debug): normalize Windows PATH env keys

### LINKED ISSUE #261: Unable to debug tests: Invalid message: Found duplicate in "env": PATH
Type: <b>Bug</b>

# Plugin Versions
Python: v2024.2.1
Python Debugger: v2024.2.0

# Behaviour
## Expected vs. Actual

Expected: When selecting "debug test" from test runner or the editor, the debugger should start.

Actual: Debugger fails to start. Popup dialog with message "Invalid message: Found duplicate in "env": PATH." is displayed.

## Steps to reproduce:

1. Open a new folder in VS Code
2. Add a single file with a dummy unit test like this:
```python
import unittest


class SomeTest(unittest.TestCase):
    def test_anything(self):
        self.assertTrue(True)
```
3. Configure Python tests
4. Try to debug the test

# Diagnostic data

-   Python version (& distribution if applicable, e.g. Anaconda): 3.8.10
-   Type of virtual environment used (e.g. conda, venv, virtualenv, etc.): Global
-   Value of the `python.languageServer` setting: Default

<details>

<summary>Output for <code>Python</code> in the <code>Output</code> panel (<code>View</code>→<code>Output</code>, change the drop-down the upper-right of the <code>Output</code> panel to <code>Python</code>)
</summary>

<p>

```
2024-03-12 17:28:25.416 [debug] Found cached env for C:\Python38\python.exe
2024-03-12 17:28:25.421 [debug] starting server as TCP
2024-03-12 17:28:25.421 [debug] starting server as TCP
2024-03-12 17:28:25.423 [info] Using launch configuration in workspace folder.
2024-03-12 17:28:25.423 [debug] Found cached env for C:\Python38\python.exe
2024-03-12 17:28:25.423 [debug] Found cached env for C:\Python38\python.exe
2024-03-12 17:28:25.464 [debug] Activation Commands received undefined for shell cmd, resource c:\UserData\z003emnm\temp\duplicate_env_repro\test1 and interpreter C:\Python38\python.exe
2024-03-12 17:28:25.472 [debug] Found cached env for C:\Python38\python.exe
2024-03-12 17:28:25.473 [debug] Found cached env for C:\Python38\python.exe
2024-03-12 17:28:25.473 [debug] Found cached env for C:\Python38\python.exe
2024-03-12 17:28:25.478 [debug] CreateEnv Trigger - turned off in settings
2024-03-12 17:28:25.616 [debug] Found cached env for C:\Python38\python.exe
2024-03-12 17:28:25.634 [debug] Extension ms-python.debugpy accessed getActiveEnvironmentPath with args: undefined
2024-03-12 17:28:25.639 [debug] Extension ms-python.debugpy accessed resolveEnvironment with args: {"id":"C:\\PYTHON38\\PYTHON.EXE","path":"C:\\Python38\\python.exe"}
2024-03-12 17:28:25.647 [debug] Found cached env for C:\Python38\python.exe
2024-03-12 17:28:25.668 [debug] Extension ms-python.debugpy accessed resolveEnvironment with args: "C:\\Python38\\python.exe"
2024-03-12 17:29:10.882 [debug] Clearing context for python dependencies not installed: ms-python.python.Python
2024-03-12 17:29:10.883 [debug] Found cached env for C:\Python38\python.exe
```

</p>
</details>

<details>
<summary>Output for Python Debugger</summary>
<p>
```
2024-03-12 17:26:32.337 [info] Name: Python Debugger
2024-03-12 17:26:32.337 [info] Module: debugpy
2024-03-12 17:27:29.664 [info] DAP Server launched with command: C:\Python38\python.exe c:\Users\z003emnm\.vscode\extensions\ms-python.debugpy-2024.2.0-win32-x64\bundled\libs\debugpy\adapter
2024-03-12 17:28:25.647 [info] DAP Server launched with command: C:\Python38\python.exe c:\Users\z003emnm\.vscode\extensions\ms-python.debugpy-2024.2.0-win32-x64\bundled\libs\debugpy\adapter
```
</p>
</details>

<details>

<summary>User Settings</summary>

<p>

```

languageServer: "Pylance"

testing
• unittestArgs: ["-v", "-s", ".", "-p", "test_*.py"]
• unittestEnabled: true

experiments
• enabled: false

```

</p>
</details>

Extension version: 2024.2.0
VS Code version: Code 1.87.0 (019f4d1419fbc8219a181fab7892ebccf7ee29a2, 2024-02-27T23:41:44.469Z)
OS version: Windows_NT x64 10.0.19045
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|11th Gen Intel(R) Core(TM) i7-11850H @ 2.50GHz (16 x 2496)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled|
|Load (avg)|undefined|
|Memory (System)|31.67GB (19.24GB free)|
|Process Argv|. --crash-reporter-id ad5562b9-4802-4e9d-83ef-990f176f3ed1|
|Screen Reader|no|
|VM|0%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vspor879:30202332
vspor708:30202333
vspor363:30204092
vscoreces:30445986
vscod805cf:30301675
binariesv615:30325510
vsaa593cf:30376535
py29gd2263:30899288
c4g48928:30535728
azure-dev_surveyone:30548225
vscrp:30673768
962ge761:30959799
pythongtdpath:30769146
welcomedialog:30910333
pythonidxpt:30866567
pythonnoceb:30805159
asynctok:30898717
pythontestfixt:30902429
pythonregdiag2:30936856
pyreplss1:30897532
pythonmypyd1:30879173
pythoncet0:30885854
pythontbext0:30879054
accentitlementsc:30887149
dsvsc016:30899300
dsvsc017:30899301
dsvsc018:30899302
pydisww1:30981707
cppperfcontrol:30979541
8082a590:30971561
fegfb526:30981948
bg6jg535:30979843
dsvsc020:30976470

```

</details>

### PR DESCRIPTION
**Pull request by xclear-cast**
## Summary
- normalize Windows search-path environment keys to the debugger's preferred Path casing before handing env vars to debugpy
- merge alternate PATH values into Path by path segment so duplicate keys and duplicate path entries are avoided
- add a VS Code-hosted unit test covering a launch env that supplies PATH while Windows uses Path

Fixes #261.

## Validation
- 
pm ci --ignore-scripts to install JS dependencies after plain 
pm ci hit a local native build-tools blocker in @vscode/windows-process-tree (MSB8040, missing Spectre-mitigated VC libraries)
- 
pm run compile-tests
- 
pm run lint
- 
pm run format-check
- 
pm run compile
- TEST_GREP='duplicate Windows search path' node ./out/test/runTest.js
- git diff --check

The focused VS Code-hosted unit test passed: 1 passing.

### CODE REVIEW

File: src/test/unittest/configuration/resolvers/helper.unit.test.ts [hunk 1]
```diff
@@ -67,4 +72,42 @@ suite('Debugging - Helpers', () => {
...
+        process.env.Path = 'C:\\Windows\\System32';
+        process.env.PATH = 'C:\\Tools';
+
+        try {
+            const launchEnv = { ['PATH']: 'C:\\Project\\.venv\\Scripts' };
+            const env = await getDebugEnvironmentVariables({
+                name: 'Python Debug Test',
+                type: 'debugpy',
+                request: 'launch',
+                env: launchEnv,
```
Comment (rchiodo) [inline src/test/unittest/configuration/resolvers/helper.unit.test.ts:90]: Copilot generated:
-95

**Test fails on non-Windows CI** (Skeptic, structurally confirmed). `sinon.stub(platform, 'getOSType').returns(platform.OSType.Windows)` makes `getSearchPathEnvVarNames()` return `['Path', 'PATH']`, but `path.delimiter` remains `:` on Linux/macOS — it's a Node.js runtime constant, not controlled by the stub. The production code in `normalizeSearchPathEnvVar` uses `path.delimiter` to split/join segments, so on Linux `env.Path` will contain `:` delimiters. The assertions then split by hardcoded `';'`, which won't match.

Fix options:
1. Use `path.delimiter` in assertions instead of hardcoded `';'`
2. Also stub/mock `path.delimiter` to `';'`
3. Skip the test on non-Windows: `if (process.platform !== 'win32') return`

[verified]

Comment (rchiodo) [inline src/test/unittest/configuration/resolvers/helper.unit.test.ts:90]: Copilot generated:
-91

**Partial assertions violate test precision** (Rules). `.to.contain(...)` would still pass if `env.Path` contained extra duplicates, wrong delimiters, or unrelated segments — exactly the regression class this test should catch:
```typescript
expect(env.Path).to.contain('C:\\Project\\.venv\\Scripts');
expect(env.Path).to.contain('C:\\Tools');
```
Fix: Capture the exact runtime value of `env.Path` (via debugger or a passing run) and assert with `.to.equal('<exact-full-string>')`. This ensures the full merged, deduplicated, correctly-delimited value is validated.

[verified]

File: src/test/unittest/configuration/resolvers/helper.unit.test.ts [hunk 2]
```diff
@@ -67,4 +72,42 @@ suite('Debugging - Helpers', () => {
 
         expect(program).to.be.equal(undefined, 'Not undefined');
     });
+
```
Comment (rchiodo) [inline src/test/unittest/configuration/resolvers/helper.unit.test.ts:75]: Copilot generated:
-102

**Missing edge case coverage** (Skeptic + Architect). The test only covers the `internalConsole` code path (which triggers `mergeVariables(process.env, env)`). Consider adding:
1. A non-`internalConsole` mode test to verify the earlier normalization calls work independently.
2. A test confirming Linux/macOS (`PATH`-only, single key) is unaffected by normalization.
3. [unverified] On Windows Node.js, `process.env` is case-insensitive, so `process.env.Path = 'X'; process.env.PATH = 'Y'` may alias to the same entry. The test's process.env setup may not create the two-key scenario it appears to — the `args.env` plain-object path still exercises the core logic, but the process.env interaction is less realistic than intended.

[verified]

File: src/extension/debugger/configuration/resolvers/helper.ts [hunk 3]
```diff
@@ -4,36 +4,76 @@
 
 'use strict';
 
+import * as path from 'path';
```
Comment (rchiodo) [inline src/extension/debugger/configuration/resolvers/helper.ts:7]: Copilot generated:
,12

**Readonly parameters** (Rules). Both new helpers accept `names: ('Path' | 'PATH')[]` but never mutate it. `vars` in `getSearchPathEnvVar` is only read (`.map`, `.find`). Per the repo's immutability-by-default convention:
```typescript
function getSearchPathEnvVar(vars: Readonly<EnvironmentVariables>, names: readonly ('Path' | 'PATH')[]): ...
function normalizeSearchPathEnvVar(env: EnvironmentVariables, names: readonly ('Path' | 'PATH')[]) { ...
```
(`env` is intentionally mutable since the function modifies it.)

[verified]

File: src/extension/debugger/configuration/resolvers/helper.ts [hunk 4]
```diff
@@ -4,36 +4,76 @@
...
+            delete env[alternatePathVariableName];
+            continue;
+        }
+
+        const currentValue = env[pathVariableName];
+        if (typeof currentValue === 'string' && currentValue.length > 0) {
+            const currentSegments = new Set(currentValue.split(path.delimiter).filter((item) => item.length > 0));
+            const alternateSegments = alternateValue
+                .split(path.delimiter)
+                .filter((item) => item.length > 0 && !currentSegments.has(item));
```
Comment (rchiodo) [inline src/extension/debugger/configuration/resolvers/helper.ts:33]: Copilot generated:
-35

**Case-sensitive path segment deduplication on a case-insensitive filesystem** (Skeptic, structurally confirmed; Architect concurs). The `Set.has()` check uses `===` semantics. On Windows, `C:\Windows\System32` and `c:\windows\system32` refer to the same directory but would survive deduplication.

The Advocate correctly notes this matches pre-existing behavior in `appendPath`/`appendPaths`. However, the PR description claims "duplicate path entries are avoided" — case-insensitive duplicates are not. Consider normalizing with `.toLowerCase()` for the lookup `Set` while preserving original casing in the output, or document this as a known limitation.

[verified]

File: src/extension/debugger/configuration/resolvers/helper.ts [hunk 5]
```diff
@@ -4,36 +4,76 @@
...
             ? ({ ...args.env } as Record<string, string>)
             : ({} as Record<string, string>);
     const envFileVars = await envParser.parseFile(args.envFile, debugLaunchEnvVars);
+    normalizeSearchPathEnvVar(debugLaunchEnvVars, pathVariableNames);
     const env = envFileVars ? { ...envFileVars } : {};
+    normalizeSearchPathEnvVar(env, pathVariableNames);
 
     // "overwrite: true" to ensure that debug-configuration env variable values
     // take precedence over env file.
     envParser.mergeVariables(debugLaunchEnvVars, env, { overwrite: true });
```
Comment (rchiodo) [inline src/extension/debugger/configuration/resolvers/helper.ts:60]: Copilot generated:
-100

**Root cause assessment — symptom treatment, not root fix** (Architect + Skeptic). Five `normalizeSearchPathEnvVar` calls are needed because `mergeVariables` in `environment.ts` only excludes `getSearchPathEnvVarNames()[0]` (i.e., `Path`), letting `PATH` through on every merge. The normalization is idempotent and correct (Advocate), but fragile — any future merge/append without a following normalize silently regresses the bug (Architect).

Recommended follow-up: fix `environment.ts` line ~55 to `settingsNotToMerge = ['PYTHONPATH', ...getSearchPathEnvVarNames()]`. This one-line change would eliminate most normalize calls and protect all consumers of `mergeVariables`, not just the debugger. If not addressed in this PR, add a TODO comment explaining why each normalize call exists.

[verified]
